src/net/neighbors/db.rs

```diff
@@ -0,0 +1,652 @@
+        let dbconn = network.peerdb_conn();
+        let mut to_resolve = vec![];
+        let mut resolved: HashMap<NeighborAddress, Neighbor> = HashMap::new();
+        for naddr in addrs {
+            let neighbor_opt =
+                Neighbor::from_neighbor_address(dbconn, network_id, block_height, naddr)?;
+
+            if let Some(neighbor) = neighbor_opt {
+                // already know about this neighbor, so look at its last contact time
+                if neighbor.last_contact_time + NEIGHBOR_MINIMUM_CONTACT_INTERVAL
+                    < get_epoch_time_secs()
+                {
+                    // stale
+                    to_resolve.push((*naddr).clone());
+                } else {
+                    // our copy is still fresh
+                    resolved.insert(naddr.clone(), neighbor);
+                }
+                continue;
+            }
+
+            // need to resolve this one, but don't talk to it if we did so recently (even
+            // if we have stale information for it -- the remote node could be trying to trick
+            // us into DDoS'ing this node).
+            let peer_opt = PeerDB::get_peer(dbconn, network_id, &naddr.addrbytes, naddr.port)
+                .map_err(net_error::DBError)?;
+
+            if let Some(n) = peer_opt {
+                // we know about this neighbor, but its key didn't match the
+                // neighboraddress.  Only try to re-connect with it if we haven't done
+                // so recently, so a rogue neighbor can't force us to DDoS another
+                // peer.
+                if n.last_contact_time + NEIGHBOR_MINIMUM_CONTACT_INTERVAL < get_epoch_time_secs() {
+                    to_resolve.push((*naddr).clone());
+                } else {
+                    // recently contacted
+                    resolved.insert(naddr.clone(), n);
+                }
+            } else {
+                // okay, we really don't know about this neighbor
+                to_resolve.push((*naddr).clone());
+            }
+        }
+        Ok((resolved, to_resolve))
+    }
+
+    /// Given a neighbor we tried to insert into the peer database, find one of the existing
+    /// neighbors it collided with.  Return its slot in the peer db.
+    fn find_replaced_neighbor_slot(
+        conn: &DBConn,
+        nk: &NeighborKey,
+    ) -> Result<Option<u32>, net_error> {
+        let mut slots = PeerDB::peer_slots(conn, nk.network_id, &nk.addrbytes, nk.port)
+            .map_err(net_error::DBError)?;
+
+        if slots.len() == 0 {
+            // not present
+            return Ok(None);
+        }
+
+        let mut rng = thread_rng();
+        slots.shuffle(&mut rng);
+
+        for slot in slots {
+            let peer_opt =
+                PeerDB::get_peer_at(conn, nk.network_id, slot).map_err(net_error::DBError)?;
+
+            match peer_opt {
+                None => {
+                    continue;
+                }
```

Comment:
Why `continue` here? If the node finds an empty slot, shouldn't it just insert there?

Reply:
Yup; I don't think this is actually reachable though.  But yes -- either way, we pick a slot at random